### PR TITLE
Initialize the CAPTCHA module with emoji CAPTCHA

### PR DESCRIPTION
**Pull request by rojvv**
- [x] Add environment variable for the API URL.
- [x] Make it only available if the environment variable is set.
- [x] Clean the filters.
- [x] Make the CAPTCHA type configurable.
- [x] ~Persist the `CryptoKey` to the database.~
- [x] Add commands to change the CAPTCHA type.
- [x] Make the message texts better.
- [x] ~Send a message in the chat once the join request is approved?~

### CODE REVIEW

File: handlers/captcha.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+ */
+
+import { Context, decrypt, encrypt } from "$utilities";
+import {
+  CallbackQueryContext,
+  Composer,
+  InlineKeyboard,
+  InputFile,
+} from "grammy";
+import { InlineKeyboardButton } from "https://deno.land/x/grammy@v1.10.1/platform.deno.ts";
```
Comment (dcdunkan) [inline handlers/captcha.ts]: grammy/types.ts?

File: handlers/captcha.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Moderent is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Afferto General Public License
+ * along with Moderent.  If not, see <https://www.gnu.org/licenses/>.
```
Comment (dcdunkan) [inline handlers/captcha.ts]: Not related to this PR ig, but there is an extra space in this line :D

File: handlers/captcha.ts (hunk 1 above)
Comment (rojvv) [reply handlers/captcha.ts]: that was auto import. will fix.

File: handlers/captcha.ts (hunk 2 above)
Comment (rojvv) [reply handlers/captcha.ts]: you mean before the "If"?

File: handlers/captcha.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      const emoji of emojis.slice(i, i + BUTTONS_PER_ROW)
+    ) {
+      let data = `emoji-captcha:${emoji}:`;
+      const available = 64 - new TextEncoder().encode(data).length;
+      data += encrypted.slice(0, available);
+      encrypted = encrypted.slice(available);
+      keyboard.text(emoji, data);
+    }
+    keyboard.row();
+  }
```
Comment (dcdunkan) [inline handlers/captcha.ts]: A suggestion (Personally I think its simpler and better than this slicing and nested loops).

https://github.com/dcdunkan/ryportbot/blob/main/bot.ts#L306-L310

File: handlers/captcha.ts (hunk 2 above)
Comment (dcdunkan) [reply handlers/captcha.ts]: Yes

Comment (rojvv) [reply handlers/captcha.ts]: It's GNU's. I'd rather leave it like that.

File: handlers/captcha.ts (hunk 3 above)
Comment (rojvv) [reply handlers/captcha.ts]: I don't know how that should be implemented here.

File: handlers/captcha/emoji.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+      const emoji of emojis.slice(i, i + BUTTONS_PER_ROW)
+    ) {
+      let data = `emoji-captcha:${emoji}:`;
+      const available = 64 - new TextEncoder().encode(data).length;
+      data += encrypted.slice(0, available);
+      encrypted = encrypted.slice(available);
+      keyboard.text(emoji, data);
+    }
+    keyboard.row();
+  }
```
Comment (dcdunkan) [inline handlers/captcha/emoji.ts]: I mean, like this:

```suggestion
  for (const [i, emoji] of emojis.entries()) {
    const emoji = emojis[i];
    let data = `emoji-captcha:${emoji}:`;
    const available = 64 - new TextEncoder().encode(data).length;
    data += encrypted.slice(0, available);
    encrypted = encrypted.slice(available);
    keyboard.text(emoji, data);
    if (i % BUTTONS_PER_ROW === (BUTTONS_PER_ROW - 1)) {
      keyboard.row();
    }
  }
```

Comment (rojvv) [reply handlers/captcha/emoji.ts]: Thanks!

**Comment by rojvv** [comment]
- [x] Don't store the AES CBC key in `localStorage` as it is not supported on Deno Deploy.

**Comment by rojvv** [comment]
Emoji CAPTCHA API spec & implementation: https://github.com/roj1512/emoji-captcha

File: database/operations/kv.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ */
+
+import { database } from "../database.ts";
+import { Collection } from "mongo";
+
+export interface Kv {
+  key: string;
+  // deno-lint-ignore no-explicit-any
+  value: any;
+}
```
Comment (ArnabXD) [inline database/operations/kv.ts]: I think having support for generic type would be great.

```suggestion
// deno-lint-ignore no-explicit-any
export interface Kv<T = any> {
  key: string;
  value: T;
}
```

Comment (rojvv) [reply database/operations/kv.ts]: Can you suggest how should the `get` and `set` functions be modified, too?

File: database/operations/kv.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        key: { "key": 1 },
+        name: "key",
+        unique: true,
+      },
+    ],
+  });
+}
+
+// deno-lint-ignore no-explicit-any
+export async function get(key: string): Promise<any> {
```
Comment (ArnabXD) [inline database/operations/kv.ts]:
```suggestion
export async function get<T = any>(key: string): Promise<T | null> {
```

File: database/operations/kv.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  let value = cache.get(key);
+  if (!value) {
+    value = (await collection.findOne({ key }))?.value ?? null;
+  }
+  return value;
+}
+
+// deno-lint-ignore no-explicit-any
+export async function set(key: string, value: any) {
```
Comment (ArnabXD) [inline database/operations/kv.ts]:
```suggestion
export async function set<T = any>(key: string, value: T) {
```

**Comment by rojvv** [comment]
lemme see

**Comment by ArnabXD** [comment]
> lemme see

Or rather `unknown` can be a better option if we want to force to pass generic while using these.

**Review by ArnabXD — APPROVED** [review]
LGTM

**Pull request merged**